### PR TITLE
e2e tests for horizontal pod autoscaling using deployments

### PR DESCRIPTION
**Pull request by xelalexv**
requested in #15474 - e2e tests now test both via replication controller and via deployment

### CODE REVIEW

**Comment by piosz** [comment]
@xelalexv thank you very much for you contribution! I'll take a look into it early next week.

@nikhiljindal could you please also take a look as the owner do Deployment feature?

**Comment by piosz** [comment]
@k8s-bot ok to test

**Comment by piosz** [comment]
The code doesn't compile with the following error:
```
output/local/go/src/k8s.io/kubernetes/test/e2e/util.go:1297: cannot use "k8s.io/kubernetes/pkg/api".PodTemplateSpec literal (type *"k8s.io/kubernetes/pkg/api".PodTemplateSpec) as type "k8s.io/kubernetes/pkg/api".PodTemplateSpec in field value
```
Could you please fix it?

File: test/e2e/horizontal_pod_autoscaling.go [hunk 1]
```diff
@@ -26,37 +27,57 @@ import (
 )
 
 const (
-	kind             = "replicationController"
+	kindRC           = "replicationController"
```
Comment (piosz) [inline test/e2e/horizontal_pod_autoscaling.go]: How about making this enum and moving to autoscaling_utils.go?

**Comment by xelalexv** [comment]
Yes, I ran the tests several times. The breaking change came with PR #16330, which was merged two days ago. I re-based & fixed accordingly.

File: test/e2e/horizontal_pod_autoscaling.go [hunk 2]
```diff
@@ -26,37 +27,57 @@ import (
...
-	// CPU tests
-	It("[Skipped][Autoscaling Suite] should scale from 1 pod to 3 pods and from 3 to 5 (scale resource: CPU)", func() {
-		rc = NewDynamicResourceConsumer("rc", 1, 250, 0, 500, 100, f)
-		defer rc.CleanUp()
-		createCPUHorizontalPodAutoscaler(rc, 20)
-		rc.WaitForReplicas(3)
-		rc.EnsureDesiredReplicas(3, stabilityTimeout)
-		rc.ConsumeCPU(700)
-		rc.WaitForReplicas(5)
+	titleUp := "[Skipped][Autoscaling Suite] should scale from 1 pod to 3 pods and from 3 to 5 (via %s, with scale resource: CPU)"
```
Comment (piosz) [inline test/e2e/horizontal_pod_autoscaling.go]: Tags [Skipped] and [Autoscaling Suite] are helpful for enabling/disabling tests in our suites. In the current situation we are unable to enable/disable test for Deployment only. Could you please move this tags to 'It' statement?

**Comment by piosz** [comment]
Overall LGTM. Just a few nits.

@nikhiljindal could you also take a look, please?

File: test/e2e/horizontal_pod_autoscaling.go (hunk 2 above)
Comment (piosz) [inline test/e2e/horizontal_pod_autoscaling.go]: Also could you please disable deployment tests by default by removing [Autoscaling Suite] tag?

Comment (xelalexv) [inline test/e2e/horizontal_pod_autoscaling.go]: Whether a test is done via RC or Deployment is currently included in the `It` statement, e.g.:
```
[Skipped][Autoscaling Suite] should scale from 1 pod to 3 pods and from 3 to 5 (via deployment, with scale resource: CPU)
[Skipped][Autoscaling Suite] should scale from 1 pod to 3 pods and from 3 to 5 (via replicationController, with scale resource: CPU)
```
So testing with deployments only could be selected with e.g. `--ginkgo.focus="Autoscaling.*deplyoment"`. 

Should that be made a dedicated `[...]` tag? E.g.
```
[Skipped][Autoscaling Suite][deployment] should scale from 1 pod to 3 pods and from 3 to 5 (with scale resource: CPU)
[Skipped][Autoscaling Suite][replicationController] should scale from 1 pod to 3 pods and from 3 to 5 (with scale resource: CPU)
```

Comment (piosz) [inline test/e2e/horizontal_pod_autoscaling.go]: Ok, let's leave it as it is now.

File: test/e2e/horizontal_pod_autoscaling.go (hunk 1 above)
Comment (piosz) [inline test/e2e/horizontal_pod_autoscaling.go]: A took a look and this is just an arbitrary string so let's not change it.

**Comment by piosz** [comment]
LGTM, could you please rebese and squash commits before merge?

**Comment by xelalexv** [comment]
Missed your comments, mea culpa! Should I revert the enum, or just do a rebase & commit squash to wrap things up?

**Comment by piosz** [comment]
Yes please revert the enum change (the last 2 commits). I though this scale subresource kind is predefined in the API but this is just an arbitrary string. Sorry for the confusion.

**Comment by piosz** [comment]
Thanks for the contribution. I'll merge it on Monday and them monitor our e2e tests.

**Pull request merged**